PWGHF/HFC/Tasks/taskCorrelationDMesonPairs.cxx

```diff
@@ -0,0 +1,569 @@
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecBkgBkg"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecBkgRef"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecBkgSig"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecRefBkg"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecRefRef"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecRefSig"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecSigBkg"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecSigRef"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtSidebandsMCRecSigSig"))->Sumw2();
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtMCGen"))->GetAxis(i)->Set(nBinspTaxis, valuespTaxis);
+      registry.get<THnSparse>(HIST("hCorrel2DVsPtMCGen"))->Sumw2();
+    }
+  }
+  ///
+  /// FUNCTIONS
+  ///
+
+  // Possible values of candidateType:
+  // |1: Sig D0              |10: Ref D0              |21: Sig D0 + Ref D0bar     |102: Bkg D0 + Sig D0bar      |201: Sig D0 + Bkg D0bar
+  // |2: Sig D0bar           |12: Ref D0 + Sig D0bar  |30: Ref D0 + Ref D0bar     |120: Bkg D0 + Ref D0bar      |210: Ref D0 + Bkg D0bar
+  // |3: Sig D0 + Sig D0bar  |20: Ref D0bar           |100: Bkg D0                |200: Bkg D0bar               |200: Bkg D0 + Bkg D0bar
+
+  // Register whether our D0 candidate is Sig, Ref or Bkg
+  uint getDMesonType(uint const& candidateType)
+  {
+    if (candidateType == 1 || candidateType == 3 || candidateType == 21 || candidateType == 201) { // Signal
+      return 1;
+    } else if (candidateType == 10 || candidateType == 12 || candidateType == 30 || candidateType == 210) { // Reflected
+      return 2;
+    } else if (candidateType == 100 || candidateType == 102 || candidateType == 120 || candidateType == 300) { // Background
+      return 3;
+    } else {
+      return 0;
+    }
+  }
+
+  // Register whether our D0bar candidate is Sig, Ref or Bkg
+  uint getDMesonBarType(uint const& candidateType)
+  {
+    if (candidateType == 2 || candidateType == 3 || candidateType == 12 || candidateType == 102) { // Signal
+      return 1;
+    } else if (candidateType == 20 || candidateType == 21 || candidateType == 30 || candidateType == 120) { // Reflected
+      return 2;
+    } else if (candidateType == 200 || candidateType == 201 || candidateType == 210 || candidateType == 300) { // Background
+      return 3;
+    } else {
+      return 0;
+    }
+  }
+
+  // Plot Mass correlations
+  void drawMassCorrPlots(uint const& candLabel1, uint const& candLabel2, double const& massCand1, double const& massCand2, double const& ptCand1, double const& ptCand2)
+  {
+    if (candLabel1 == 1 && candLabel2 == 1) {
+      registry.fill(HIST("hMass2DCorrelationPairsMCRecSigSig"), massCand1, massCand2, ptCand1, ptCand2);
+    } else if (candLabel1 == 1 && candLabel2 == 2) {
```

Comment:
Use a 2D array of histogram pointers and directly fill the right histogram using indices.